lib/client-retry.js

```diff
@@ -0,0 +1,170 @@
+
+        }
+        // pass through error
+        if (error) {
+            throw error;
+        }
+        return response;
+    }
+
+    // retry on 429s
+    async unregister() {
+        let response, error;
+        let non429Response = false;
+        let retryCount = 0;
+        while (this.shouldRetry(non429Response, retryCount)) {
+            retryCount++;
+            try {
+                response = await super.unregister();
+                non429Response = true;
+            } catch (e) {
+                // string comparision instead of `instanceof` since `instanceof` causes issues
+                if (e.code === TOO_MANY_REQUEST_ERROR_CODE && e.name === TOO_MANY_REQUEST_ERROR) {
+                    non429Response = false;
+                    console.log(colors.red(`encountered a 429: ${e.message}.`));
+                    const waitTime = this.retryWaitTime(e.retryAfter);
+                    console.log(colors.red(`Waiting for ${waitTime}ms and trying again`));
+                    await sleep(waitTime);
+                } else {
+                    non429Response = true;
+                    error = e;
+                }
+            }
+        }
+
+        if (!non429Response) {
+            throw new Error(`Running into 429s after ${retryCount} attempts. Will not continue retrying. Try again in a few minutes.`);
+
+        }
+        // pass through error
+        if (error) {
+            throw error;
+        }
+        return response;
+    }
+
+    async process(source, renditions, userData) {
+        let response, error;
+        let non429Response = false;
+        let retryCount = 0;
+        while (this.shouldRetry(non429Response, retryCount)) {
+            retryCount++;
+            try {
+                response = await super.process(source, renditions, userData);
+                console.log('Successful call to /process', retryCount);
+                non429Response = true;
+            } catch (e) {
+                // string comparision instead of `instanceof` since `instanceof` causes issues
+                if (e.code === TOO_MANY_REQUEST_ERROR_CODE && e.name === TOO_MANY_REQUEST_ERROR) {
+                    non429Response = false;
+                    console.log(colors.red(`encountered a 429: ${e.message}.`));
+                    const waitTime = this.retryWaitTime(e.retryAfter);
+                    console.log(colors.red(`Waiting for ${waitTime}ms and trying again`));
+                    await sleep(waitTime);
+                } else {
+                    non429Response = true;
+                    error = e;
+                }
+            }
+        }
+
+        if (!non429Response) {
+            throw new Error(`Running into 429s after ${retryCount} attempts. Will not continue retrying. Try again in a few minutes.`);
+
+        }
+        // pass through error
+        if (error) {
+            throw error;
+        }
+        return response;
+    }
```

Comment:
Maybe it would be easier to create a retry function to contain the retry behavior so the duplication can be avoided. The function could be something like:

```
async function retryOnTooManyRequests(func) {
   //... retry behavior
   const response = await func();
   //... retry behavior
}
```

I did this here for http upload/download: https://github.com/adobe/node-httptransfer/blob/master/lib/retry.js#L169

The options passing is a bit messy, but adding retry is pretty straight forward: https://github.com/adobe/node-httptransfer/blob/master/lib/file.js#L103-L114

Reply:
I took inspiration from you retry function in node-httptransfer and added something similar here: https://github.com/adobe/asset-compute-client/blob/NUI-1650/lib/retry.js

Now instead of a separate class, the 429 retry logic is integrated into `assetcompute.js`: https://github.com/adobe/asset-compute-client/blob/NUI-1650/lib/assetcompute.js#L93-L94

However, `assetcompute.js` already uses `node-fetch-retry` for retry logic. So, I kept the `retry` function only for custom 429 retry logic. and we will use `node-fetch-retry` for the rest of the retry logic: https://github.com/adobe/asset-compute-client/blob/NUI-1650/lib/assetcompute.js#L15